contracts/src/FunctionGateway.sol

```diff
@@ -2,246 +2,295 @@
 
-    /// @dev Keeps track of the nonce for generating request ids.
-    uint256 public nonce;
+    /// @dev The currently verified function identifier.
+    bytes32 public verifiedFunctionId;
 
-    /// @dev Maps request ids to their corresponding requests.
-    mapping(bytes32 => FunctionRequest) public requests;
+    /// @dev The currently verified function input hash.
+    bytes32 public verifiedInputHash;
 
-    /// @notice The dynamic scalar for requests.
-    uint256 public scalar;
+    /// @dev The currently verified function output.
+    bytes public verifiedOutput;
 
-    /// @notice A reference to the contract where fees are sent.
-    /// @dev During the request functions, this is used to add msg.value to the sender's balance.
-    address public feeVault;
+    /// @dev A flag that indicates whether the contract is currently making a callback.
+    bool public isCallback;
 
-    function initialize(uint256 _scalar, address _feeVault, address _timelock, address _guardian)
-        external
-        initializer
-    {
-        scalar = _scalar;
-        feeVault = _feeVault;
+    /// @dev Initializes the contract.
+    /// @param _timelock The address of the timelock contract.
+    /// @param _guardian The address of the guardian.
+    function initialize(
+        address _timelock,
+        address _guardian
+    ) external initializer {
+        isCallback = false;
         __TimelockedUpgradeable_init(_timelock, _guardian);
     }
 
-    function request(bytes32 _functionId, bytes memory _input, bytes4 _callbackSelector, bytes memory _context)
-        external
-        payable
-        returns (bytes32)
-    {
-        return request(_functionId, _input, _callbackSelector, _context, DEFAULT_GAS_LIMIT, tx.origin);
-    }
-
-    /// @dev Requests for a proof to be generated by the marketplace.
-    /// @param _functionId The id of the proof to be generated.
-    /// @param _input The input to the proof.
-    /// @param _context The context of the runtime.
+    /// @dev Creates a onchain request for a proof. The output and proof is fulfilled asynchronously
+    ///      by the provided callback.
+    /// @param _functionId The function identifier.
+    /// @param _input The function input.
+    /// @param _context The function context.
     /// @param _callbackSelector The selector of the callback function.
-    function request(
+    /// @param _callbackGasLimit The gas limit for the callback function.
+    function requestCallback(
         bytes32 _functionId,
         bytes memory _input,
-        bytes4 _callbackSelector,
         bytes memory _context,
-        uint256 _gasLimit,
-        address _refundAccount
-    ) public payable returns (bytes32) {
+        bytes4 _callbackSelector,
+        uint32 _callbackGasLimit
+    ) external payable returns (bytes32) {
+        // Compute the callback hash uniquely associated with this request.
         bytes32 inputHash = sha256(_input);
```

Comment:
why are you hashing here? you should just shove it all into the _requestHash keccack, just shove all the bytes including `_input` and `context` in there

Reply:
actually i understand--it's so that `_input` and `context` don't conflict i guess since they're both unsized, nvrmind